src/VisualStudio/Core/Def/Implementation/FindReferences/Filters/ExternalSourcesFilterHandlerBase.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Internal.VisualStudio.Shell.ErrorList;
+using Microsoft.VisualStudio.Shell.TableControl;
+using Microsoft.VisualStudio.Shell.TableManager;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.FindReferences.Filters
+{
+    internal abstract class ExternalSourcesFilterHandlerBase : FilterHandlerBase
+    {
+        public abstract bool IncludeExact { get; }
+        public abstract bool IncludeExactMetadata { get; }
+        public abstract bool IncludeOther { get; }
+
+        public override IEntryFilter GetFilter(out string displayText)
+        {
+            displayText = FilterDisplayName;
+            return new EntryFilter(IncludeExact, IncludeExactMetadata, IncludeOther);
+        }
+
+        private class EntryFilter : IEntryFilter
+        {
+            private readonly bool _includeExact;
+            private readonly bool _includeExactMetadata;
+            private readonly bool _includeOther;
+
+            public EntryFilter(bool includeExact, bool includeExactMetadata, bool includeOther)
+            {
+                _includeExact = includeExact;
+                _includeExactMetadata = includeExactMetadata;
+                _includeOther = includeOther;
+            }
+
+            public bool Match(ITableEntryHandle entry)
+            {
+                if (!entry.TryGetValue(StandardTableKeyNames.ItemOrigin, out ItemOrigin origin))
+                    origin = ItemOrigin.Exact;
```

Comment:
what is an ItemOrigin?

Reply:
https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.tablemanager.itemorigin?view=visualstudiosdk-2022